__Steps To Preform Data Preprocessing__ 

● Removing duplicate rows.

● Removing null values from rows.

● Export the cleaned dataset as a .csv file: prefer UTF-8 encoding.

● Convert the pre-processed dataset into an SQL file. 

● Manually generate a table by utilizing the database information provided in the "Database Info" tab.

In [1]:
#Import the Pandas library for Data Pre-processing

import pandas as pd

__Import the Data__

In [2]:
df = pd.read_csv('C:\\Users\\Ibrahim\\Dropbox\\PC\\Downloads\\Uncleaned_employees_final_dataset.csv')
df

employee_id         department     region  education gender  \
0             8724         Technology  region_26  Bachelors      m   
1            74430                 HR   region_4  Bachelors      f   
2            72255  Sales & Marketing  region_13  Bachelors      m   
3            38562        Procurement   region_2  Bachelors      f   
4            64486            Finance  region_29  Bachelors      m   
...            ...                ...        ...        ...    ...   
17412        64573         Technology   region_7  Bachelors      f   
17413        49584                 HR   region_7  Bachelors      m   
17414        49584                 HR   region_7  Bachelors      m   
17415        49584                 HR   region_7  Bachelors      m   
17416        64573                 HR   region_7  Bachelors      f   

      recruitment_channel  no_of_trainings  age  previous_year_rating  \
0                sourcing                1   24                   NaN   
1                   other                1   31                   3.0   
2                   other                1   31                   1.0   
3                   other                3   31                   2.0   
4                sourcing                1   30                   4.0   
...                   ...              ...  ...                   ...   
17412            referred                2   30                   5.0   
17413               other                1   33                   1.0   
17414               other                1   33                   1.0   
17415               other                1   33                   1.0   
17416            referred                1   35                   5.0   

       length_of_service  KPIs_met_more_than_80  awards_won  \
0                      1                      1           0   
1                      5                      0           0   
2                      4                      0           0   
3                      9                      0           0   
4                      7                      0           0   
...                  ...                    ...         ...   
17412                  6                      1           0   
17413                  9                      0           0   
17414                  9                      0           0   
17415                  9                      0           0   
17416                  4                      1           0   

       avg_training_score  
0                      77  
1                      51  
2                      47  
3                      65  
4                      61  
...                   ...  
17412                  81  
17413                  51  
17414                  51  
17415                  51  
17416                  81  

[17417 rows x 13 columns]

Understanding how is the data structure

In [3]:
df.describe()

employee_id  no_of_trainings           age  previous_year_rating  \
count  17417.000000     17417.000000  17417.000000          16054.000000   
mean   39083.491129         1.250732     34.807774              3.345459   
std    22707.024087         0.595692      7.694046              1.265386   
min        3.000000         1.000000     20.000000              1.000000   
25%    19281.000000         1.000000     29.000000              3.000000   
50%    39122.000000         1.000000     33.000000              3.000000   
75%    58838.000000         1.000000     39.000000              4.000000   
max    78295.000000         9.000000     60.000000              5.000000   

       length_of_service  KPIs_met_more_than_80    awards_won  \
count       17417.000000           17417.000000  17417.000000   
mean            5.801860               0.358845      0.023368   
std             4.175533               0.479675      0.151074   
min             1.000000               0.000000      0.000000   
25%             3.000000               0.000000      0.000000   
50%             5.000000               0.000000      0.000000   
75%             7.000000               1.000000      0.000000   
max            34.000000               1.000000      1.000000   

       avg_training_score  
count        17417.000000  
mean            63.176322  
std             13.418179  
min             39.000000  
25%             51.000000  
50%             60.000000  
75%             75.000000  
max             99.000000

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17417 entries, 0 to 17416
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   employee_id            17417 non-null  int64  
 1   department             17417 non-null  object 
 2   region                 17417 non-null  object 
 3   education              16646 non-null  object 
 4   gender                 17417 non-null  object 
 5   recruitment_channel    17417 non-null  object 
 6   no_of_trainings        17417 non-null  int64  
 7   age                    17417 non-null  int64  
 8   previous_year_rating   16054 non-null  float64
 9   length_of_service      17417 non-null  int64  
 10  KPIs_met_more_than_80  17417 non-null  int64  
 11  awards_won             17417 non-null  int64  
 12  avg_training_score     17417 non-null  int64  
dtypes: float64(1), int64(7), object(5)
memory usage: 1.7+ MB


In [11]:
df.isnull().sum()

employee_id                 0
department                  0
region                      0
education                 771
gender                      0
recruitment_channel         0
no_of_trainings             0
age                         0
previous_year_rating     1363
length_of_service           0
KPIs_met_more_than_80       0
awards_won                  0
avg_training_score          0
dtype: int64

__Removing duplicate rows__

In [14]:
df=df.drop_duplicates()

In [15]:
df

employee_id         department     region  education gender  \
0             8724         Technology  region_26  Bachelors      m   
1            74430                 HR   region_4  Bachelors      f   
2            72255  Sales & Marketing  region_13  Bachelors      m   
3            38562        Procurement   region_2  Bachelors      f   
4            64486            Finance  region_29  Bachelors      m   
...            ...                ...        ...        ...    ...   
17410        57239  Sales & Marketing  region_31  Bachelors      m   
17411        73858  Sales & Marketing  region_25  Bachelors      m   
17412        64573         Technology   region_7  Bachelors      f   
17413        49584                 HR   region_7  Bachelors      m   
17416        64573                 HR   region_7  Bachelors      f   

      recruitment_channel  no_of_trainings  age  previous_year_rating  \
0                sourcing                1   24                   NaN   
1                   other                1   31                   3.0   
2                   other                1   31                   1.0   
3                   other                3   31                   2.0   
4                sourcing                1   30                   4.0   
...                   ...              ...  ...                   ...   
17410               other                1   42                   3.0   
17411            sourcing                2   28                   2.0   
17412            referred                2   30                   5.0   
17413               other                1   33                   1.0   
17416            referred                1   35                   5.0   

       length_of_service  KPIs_met_more_than_80  awards_won  \
0                      1                      1           0   
1                      5                      0           0   
2                      4                      0           0   
3                      9                      0           0   
4                      7                      0           0   
...                  ...                    ...         ...   
17410                 10                      0           0   
17411                  5                      0           0   
17412                  6                      1           0   
17413                  9                      0           0   
17416                  4                      1           0   

       avg_training_score  
0                      77  
1                      51  
2                      47  
3                      65  
4                      61  
...                   ...  
17410                  53  
17411                  45  
17412                  81  
17413                  51  
17416                  81  

[17415 rows x 13 columns]

__Removing Null values from rows__

In [16]:
df=df.dropna()

In [17]:
df

employee_id         department     region  education gender  \
1            74430                 HR   region_4  Bachelors      f   
2            72255  Sales & Marketing  region_13  Bachelors      m   
3            38562        Procurement   region_2  Bachelors      f   
4            64486            Finance  region_29  Bachelors      m   
5            46232        Procurement   region_7  Bachelors      m   
...            ...                ...        ...        ...    ...   
17410        57239  Sales & Marketing  region_31  Bachelors      m   
17411        73858  Sales & Marketing  region_25  Bachelors      m   
17412        64573         Technology   region_7  Bachelors      f   
17413        49584                 HR   region_7  Bachelors      m   
17416        64573                 HR   region_7  Bachelors      f   

      recruitment_channel  no_of_trainings  age  previous_year_rating  \
1                   other                1   31                   3.0   
2                   other                1   31                   1.0   
3                   other                3   31                   2.0   
4                sourcing                1   30                   4.0   
5                sourcing                1   36                   3.0   
...                   ...              ...  ...                   ...   
17410               other                1   42                   3.0   
17411            sourcing                2   28                   2.0   
17412            referred                2   30                   5.0   
17413               other                1   33                   1.0   
17416            referred                1   35                   5.0   

       length_of_service  KPIs_met_more_than_80  awards_won  \
1                      5                      0           0   
2                      4                      0           0   
3                      9                      0           0   
4                      7                      0           0   
5                      2                      0           0   
...                  ...                    ...         ...   
17410                 10                      0           0   
17411                  5                      0           0   
17412                  6                      1           0   
17413                  9                      0           0   
17416                  4                      1           0   

       avg_training_score  
1                      51  
2                      47  
3                      65  
4                      61  
5                      68  
...                   ...  
17410                  53  
17411                  45  
17412                  81  
17413                  51  
17416                  81  

[15422 rows x 13 columns]

In [18]:
# checking is there any null values

df.isnull().sum()

employee_id              0
department               0
region                   0
education                0
gender                   0
recruitment_channel      0
no_of_trainings          0
age                      0
previous_year_rating     0
length_of_service        0
KPIs_met_more_than_80    0
awards_won               0
avg_training_score       0
dtype: int64

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 15422 entries, 1 to 17416
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   employee_id            15422 non-null  int64  
 1   department             15422 non-null  object 
 2   region                 15422 non-null  object 
 3   education              15422 non-null  object 
 4   gender                 15422 non-null  object 
 5   recruitment_channel    15422 non-null  object 
 6   no_of_trainings        15422 non-null  int64  
 7   age                    15422 non-null  int64  
 8   previous_year_rating   15422 non-null  float64
 9   length_of_service      15422 non-null  int64  
 10  KPIs_met_more_than_80  15422 non-null  int64  
 11  awards_won             15422 non-null  int64  
 12  avg_training_score     15422 non-null  int64  
dtypes: float64(1), int64(7), object(5)
memory usage: 1.6+ MB


After Pre-processed the data we have total 15422 rows and 13 columns
↓↓

In [22]:
df.shape

(15422, 13)

Then Export the clean data for further analysis in SQL

In [33]:
dataset=df.to_csv('C:\\Users\\Ibrahim\\Dropbox\\PC\\Downloads\\clean_data.csv')

After export pre-processed dataset we should import into SQL server,

then we will be working on performing data analysis using SQL on the pre-processed.